src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs

```diff
@@ -909,9 +917,10 @@ private BoundExpression CreateAndPopulateList(BoundCollectionExpression node, Ty
                 Debug.Assert(useKnownLength);
                 Debug.Assert(setCount is { });
                 Debug.Assert(asSpan is { });
+                Debug.Assert(knownLengthTemp is { });
```

Comment:
This expectation may not be met if the definition of List lacks a constructor with `int capacity`. Since pretty much every real definition of it has that constructor, it would probably be fine to just disable optimizations in that case. We should be able to test that case by calling `compilation.MakeMemberMissing` from the tests.

Reply:
Today if `CollectionsMarshal.SetCount` and `CollectionsMarshal.AsSpan` don't exist, then optimizations are disabled, and the code assumes that `List.ctor(int32)` exists. There's nothing in place that checks that `List.ctor(int32)` exists, so the assumption is already baked in, no? So today, if you're compiling a collection expression on a platform that doesn't have CollectionMarshals but also doesn't have List.ctor(int32), we're in troubles.

I can certainly fix this as part of this PR, but that's not where the assert you highlighted may not be met, I think.

Taking another look at the code, the problem arises when:

1. We satisfy all the requirements to enter `useOptimizations = true;`
2. We do not enter the code path to use the `List.ctor(in32)`, so `knownLengthTemp` is not initialized.

That should only happen if `_factory.CurrentFunction?.IsAsync == true`

I'll push a change that makes the assumptions and assertions clearer: if we set `useOptimizations`, then the expectation is that `knownLengthTemp` is initialized.

Reply:
I'll add a test for the case where optimizations are disabled because we're in an async method.

Reply:
Ah, got it, looks like compiler emits an error when `List(int)` is missing, we don't need to account for its absence in lowering.